tempesta_fw/http.c

```diff
@@ -1494,19 +1520,9 @@ tfw_http_conn_cli_drop(TfwCliConn *cli_conn)
 	 * condition with freeing of a request in tfw_http_resp_fwd().
 	 */
 	spin_lock(&cli_conn->seq_qlock);
-	list_for_each_entry_safe(req, tmp, seq_queue, msg.seq_list) {
-		if (req->resp)
-			list_move_tail(&req->msg.seq_list, &zap_queue);
-		else
-			list_del_init(&req->msg.seq_list);
-	}
+	list_for_each_entry_safe(req, tmp, seq_queue, msg.seq_list)
+		list_del_init(&req->msg.seq_list);
 	spin_unlock(&cli_conn->seq_qlock);
-
-	list_for_each_entry_safe(req, tmp, &zap_queue, msg.seq_list) {
-		BUG_ON(!list_empty_careful(&req->fwd_list));
-		BUG_ON(!list_empty_careful(&req->nip_list));
-		__tfw_http_resp_pair_free(req);
-	}
```

Comment:
A client sends us several requests and the client connection immediately terminated. All of the requests are in `fwd_queue`'s and are being sent to servers, reforwarded and so on until all of the receive responses. Then `tfw_http_resp_fwd()` finds empty `seq_queue` of the client connection and calls `ss_close_sync()` for each of the requests. All in all, we do a lot of unnecessary work.

It's OK to leave the dead requests in `fwd_queue`'s to minimize lock contention, but it's easy and cheap to:
1. set some flag for the reuests to make them evicted from `fwd_queue` right they're going to be send to a server or rescheduled;
2. we don't need to call `ss_close_sync()` in `tfw_http_resp_fwd()` since all `connection_error` and `connection_drop` callbacks follow `ss_do_close()` call in sock.c, so `tfw_http_conn_cli_drop()` is always called on closed socket.
3. `tfw_http_conn_cli_drop()` should call `tfw_connection_put()` just like `tfw_sock_srv_connect_drop()` and `tfw_sock_srv_connect_failover()` do. When `tfw_http_resp_fwd()` calls `tfw_http_conn_msg_free()` for the last request referencing the client connection, the connection is freed.

Reply:
> we don't need to call ss_close_sync() in tfw_http_resp_fwd() since all connection_error and connection_drop callbacks follow ss_do_close() call in sock.c, so tfw_http_conn_cli_drop() is always called on closed socket.

I'm agree with the comment: https://github.com/tempesta-tech/tempesta/blob/f287f9f910131f576d66b9e3f3884537e92eff04/tempesta_fw/http.c#L2089-L2091
Need fo close client connection if it's open and `seq_queue` is empty.

> tfw_http_conn_cli_drop() should call tfw_connection_put() just like tfw_sock_srv_connect_drop() and tfw_sock_srv_connect_failover() do. 

`tfw_http_conn_cli_drop()` is a `TfwConnHooks`, while both
 `tfw_sock_srv_connect_drop()` and `tfw_sock_srv_connect_failover()` are `SsHooks`. Not  `tfw_http_conn_cli_drop()`  but `tfw_sock_clnt_drop()` and `tfw_sock_clnt_error()` must put the client connection. It's already done this way, so nothing to do at this point.